pkg/codegen/go/gen_program_expressions.go

```diff
@@ -432,11 +617,24 @@ func (nameInfo) Format(name string) string {
 }
 
 // lowerExpression amends the expression with intrinsics for C# generation.
-func (g *generator) lowerExpression(expr model.Expression, typ model.Type) (model.Expression, []*ternaryTemp) {
+func (g *generator) lowerExpression(expr model.Expression, typ model.Type, isInput bool) (
+	model.Expression, []interface{}) {
 	expr, diags := hcl2.RewriteApplies(expr, nameInfo(0), false /*TODO*/)
 	expr = hcl2.RewriteConversions(expr, typ)
-	expr, temps, ternDiags := g.rewriteTernaries(expr)
+	expr, tTemps, ternDiags := g.rewriteTernaries(expr, g.ternaryTempSpiller)
+	expr, jTemps, jsonDiags := g.rewriteToJSON(expr, g.jsonTempSpiller)
```

Comment:
These should be run prior to the apply rewriter, otherwise there is a risk that the expressions that are being lifted will be lifted out of an `apply` callback.

Reply:
Eventually, the value expressions for these temps should be passed through `lowerExpression` as well (this will ensure that they pick up the `apply` transform et. al.). No need to fix now, though.